pkg/util/limiter/query_limiter_test.go

```diff
@@ -69,26 +95,97 @@ func TestQueryLimiter_AddSeries_ShouldReturnErrorOnLimitExceeded(t *testing.T) {
+	memoryTracker := NoopMemoryTracker{}
+	returnedSeries1, err := limiter.AddSeries(series1, memoryTracker)
 	require.NoError(t, err)
+	assertSameLabels(t, returnedSeries1, series1)
 	assertRejectedQueriesMetricValue(t, reg, 0, 0, 0, 0)
 
-	err = limiter.AddSeries(series2)
+	returnedSeries2, err := limiter.AddSeries(series2, memoryTracker)
 	require.Error(t, err)
+	// When limit is exceeded, empty labels are returned
+	require.True(t, returnedSeries2.IsEmpty(), "should return empty labels on error")
 	assertRejectedQueriesMetricValue(t, reg, 1, 0, 0, 0)
 
 	// Add the same series again and ensure that we don't increment the failed queries metric again.
-	err = limiter.AddSeries(series2)
+	returnedSeries2Again, err := limiter.AddSeries(series2, memoryTracker)
 	require.Error(t, err)
+	require.True(t, returnedSeries2Again.IsEmpty(), "should return empty labels on error")
 	assertRejectedQueriesMetricValue(t, reg, 1, 0, 0, 0)
 
 	// Add another series and ensure that we don't increment the failed queries metric again.
-	err = limiter.AddSeries(series3)
+	returnedSeries3, err := limiter.AddSeries(series3, memoryTracker)
 	require.Error(t, err)
+	require.True(t, returnedSeries3.IsEmpty(), "should return empty labels on error")
 	assertRejectedQueriesMetricValue(t, reg, 1, 0, 0, 0)
+
+	// Add another series with duplicate labels and ensure that we don't increment the failed queries metric again.
+	returnedSeries3NewInstance, err := limiter.AddSeries(series3NewInstance, memoryTracker)
+	require.Error(t, err)
+	require.True(t, returnedSeries3NewInstance.IsEmpty(), "should return empty labels on error")
+	assertRejectedQueriesMetricValue(t, reg, 1, 0, 0, 0)
+}
+
+func TestQueryLimiter_AddSeries_HashCollision(t *testing.T) {
+	const metricName = "test_metric"
+
+	// Create two different series
+	series1 := labels.FromMap(map[string]string{
+		model.MetricNameLabel: metricName + "_1",
+		"label1":              "value1",
+	})
+	series2 := labels.FromMap(map[string]string{
+		model.MetricNameLabel: metricName + "_2",
+		"label2":              "value2",
+	})
+	series3 := labels.FromMap(map[string]string{
+		model.MetricNameLabel: metricName + "_3",
+		"label3":              "value3",
+	})
+
+	reg := prometheus.NewPedanticRegistry()
+	limiter := NewQueryLimiter(100, 0, 0, 0, stats.NewQueryMetrics(reg))
+
+	// Test scenario: Manually simulate a hash collision state
+	// In reality, hash collisions are extremely rare with 64-bit hashes,
+	// but we need to verify the code handles them correctly.
+
+	// Use a common hash for series1 and series2 to simulate collision
+	collisionHash := uint64(12345)
+
+	// Manually set up the collision state in the limiter
+	limiter.conflictSeries = make(map[uint64][]labels.Labels)
+	// The first series is in uniqueSeries map.
+	limiter.uniqueSeries[collisionHash] = series1
+	// If another series hash is colliding, AddSeries will add it to conflictSeries map.
+	limiter.conflictSeries[collisionHash] = []labels.Labels{series2}
```

Comment:
Why do we need to do this? Couldn't we call `limiter.AddSeries(series1)`, `limiter.AddSeries(series2)` etc.?